<a href="https://colab.research.google.com/github/nuhaaaaan/Machine-Learning-Assignments-/blob/main/Exercise_3_SVM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="../src/packt-banner.png" alt="">

### Exercise 3: Build a SVM modle for Face Recognition Problem
##### (25 points) --> your total will divided by 5 to get 5 points for this exercise.
---

We will use a very famous dataset, called Labelled Faces in the Wild, which
consists of 1288 faces of famous people, and it is available at http://viswww.cs.umass.edu/lfw/lfw-funneled.tgz.

However, note that it can be easily imported via scikit-learn from the datasets class.
Each image consists of 1850 features: we could proceed by simply using each of them in the model.



Fitting a SVM to non-linear data using the Kernel Trick produces non- linear decision boundaries.
In particular, we seek to:
* Build SVM model with radial basis function (RBF) kernel
* Use a grid search cross-validation to explore ran- dom combinations of parameters.

### Step to do:

1. Loading the dataf from sklearn.datasets:

In [ ]:
from sklearn.datasets import fetch_lfw_people
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, accuracy_score
faces = fetch_lfw_people(min_faces_per_person=60)

2. Since the data can be accessed from the sklearn.datasets module, you need to explore the dataset.
    - (refer to the first 6 steps in Lab_1 could help you)

a- Print the field names (that is, the keys to the dictionary) (1 point)

In [ ]:
# What fields are in the dictionary?
print(faces.keys())

dict_keys(['data', 'images', 'target', 'target_names', 'DESCR'])


b- Print the dataset description contained (2 point)

In [ ]:
# write your code here
print(faces.DESCR)

.. _labeled_faces_in_the_wild_dataset:

The Labeled Faces in the Wild face recognition dataset
------------------------------------------------------

This dataset is a collection of JPEG pictures of famous people collected
over the internet, all details are available on the official website:

http://vis-www.cs.umass.edu/lfw/

Each picture is centered on a single face. The typical task is called
Face Verification: given a pair of two pictures, a binary classifier
must predict whether the two images are from the same person.

An alternative task, Face Recognition or Face Identification is:
given the picture of the face of an unknown person, identify the name
of the person by referring to a gallery of previously seen pictures of
identified persons.

Both Face Verification and Face Recognition are tasks that are typically
performed on the output of a model trained to perform Face Detection. The
most popular model for Face Detection is called Viola-Jones and is
implemented in the OpenCV li

3. Print the data, its shape, and the target names. ( 3 points)

In [ ]:
# What does the data look like?
print(faces.data)

[[0.53464055 0.5254902  0.49673203 ... 0.00653595 0.00653595 0.00261438]
 [0.28627452 0.20784314 0.2522876  ... 0.96993464 0.9490196  0.9346406 ]
 [0.31895426 0.39215687 0.275817   ... 0.4261438  0.7908497  0.9555555 ]
 ...
 [0.11633987 0.11111111 0.10196079 ... 0.5686274  0.5803922  0.5542484 ]
 [0.19346406 0.21176471 0.2901961  ... 0.6862745  0.654902   0.5908497 ]
 [0.12287582 0.09803922 0.10980392 ... 0.12941177 0.1633987  0.29150328]]


In [ ]:
# what is the shape of the data?
print(faces.data.shape)

(1348, 2914)


In [ ]:
# What is the target names?
print(faces.target_names)

['Ariel Sharon' 'Colin Powell' 'Donald Rumsfeld' 'George W Bush'
 'Gerhard Schroeder' 'Hugo Chavez' 'Junichiro Koizumi' 'Tony Blair']


4. Divide the data into features (X) using the faces.data and target (y) using faces.target (2 points)

In [ ]:
# Write your code here
X =faces.data
y =faces.target

5. Splitting the data into training and testing sets. (2 point)

We train the model with 70% of the samples and test with the remaining 30%.

In [ ]:
# Write your code here
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)




# print the sizes of our training and test set to verify if the splitting has occurred properly.
print(f"Training set size: {X_train.shape}, Testing set size: {X_test.shape}")


Training set size: (943, 2914), Testing set size: (405, 2914)


6. Declare SVM model with kernel='rbf', class_weight='balanced' (2 points)

In [ ]:
from sklearn.svm import SVC

# write your code here
svm_model = SVC(kernel='rbf', class_weight='balanced')

7. Use a grid search cross-validationwith 10 CV to explore random combinations of parameters. (3 points)
    - we will adjust C, which controls the margin
    - and Gamma (γ), which controls the size of the radial basis function kernel, and determine the best model.

In [ ]:
from sklearn.model_selection import GridSearchCV

# Define the parameter grid
param_grid = {'C': [1, 5, 10, 50], 'gamma': [0.001, 0.0005, 0.01, 0.1]}

# Grid search with 10-fold cross-validation
grid_search = GridSearchCV(svm_model, param_grid, cv=10)
grid_search.fit(X_train, y_train)

# Print the best parameters found by the grid search
print(f'Best parameters: {grid_search.best_params_}')

Best parameters: {'C': 50, 'gamma': 0.001}


8. predict on the test set, using the best model from above step (best_estimator_) (5 points)

In [ ]:
# write your code here
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

# Print the predictions
print(f'Predictions on the test set: {y_pred}')

Predictions on the test set: [3 4 3 6 6 1 3 3 3 3 3 3 4 3 3 1 7 2 3 2 7 7 5 5 0 3 6 7 3 3 0 6 3 3 2 3 2
 3 3 2 3 3 7 1 3 3 0 2 1 2 7 3 4 6 7 3 7 1 7 0 4 2 7 2 5 4 7 3 4 3 1 3 4 1
 3 4 0 4 3 3 1 3 1 0 7 2 3 2 7 0 1 1 2 3 3 1 7 3 3 1 3 7 1 4 3 3 0 3 7 3 3
 1 0 1 3 1 3 2 3 4 7 7 5 4 3 3 3 3 2 2 3 3 0 3 4 3 4 1 2 1 7 6 5 3 3 1 0 3
 4 4 3 2 7 1 7 1 3 7 1 4 6 1 2 3 2 3 1 7 2 2 1 7 3 3 1 1 1 3 3 1 0 0 1 1 7
 1 1 5 3 4 3 3 7 5 6 3 7 7 3 2 2 3 2 3 3 6 3 3 1 7 3 6 1 3 3 1 1 7 6 3 1 3
 1 7 7 2 7 7 5 7 1 3 3 2 3 4 7 2 3 1 3 7 7 1 4 3 1 1 5 4 2 3 4 1 1 1 2 2 3
 7 3 7 3 7 3 1 3 1 3 1 1 1 3 3 1 4 4 3 1 4 1 0 0 3 2 0 2 5 1 3 3 6 2 1 3 6
 3 1 3 5 1 1 1 3 3 3 3 3 3 5 0 1 1 3 3 1 1 7 1 1 6 3 1 1 5 7 3 2 1 3 1 3 3
 7 3 1 0 3 6 3 7 3 3 0 5 3 1 6 3 0 1 1 0 1 0 4 1 3 2 2 3 1 4 4 2 3 1 3 1 1
 3 1 3 2 3 2 7 1 7 3 1 2 6 4 1 7 3 3 3 3 3 2 3 3 5 1 1 2 5 6 2 5 7 3 3]


9. Model performances:
Run the following code to print the model evaluation metric

In [ ]:
from sklearn.metrics import classification_report
labels = list(faces.target_names)
print(classification_report(y_test,y_pred,target_names=labels))

                   precision    recall  f1-score   support

     Ariel Sharon       0.59      0.76      0.67        17
     Colin Powell       0.80      0.83      0.81        84
  Donald Rumsfeld       0.67      0.81      0.73        36
    George W Bush       0.90      0.86      0.88       146
Gerhard Schroeder       0.69      0.71      0.70        28
      Hugo Chavez       1.00      0.63      0.77        27
Junichiro Koizumi       0.89      1.00      0.94        16
       Tony Blair       0.82      0.78      0.80        51

         accuracy                           0.81       405
        macro avg       0.79      0.80      0.79       405
     weighted avg       0.83      0.81      0.82       405



10. What do you observe about the model performances? (5 points)

#### Write your answer here

Observations on Model Performance:
Noteworthy Accuracy:

The model achieves an impressive overall accuracy of 81%, indicating strong performance in the face recognition task. This result reflects the effectiveness of using an SVM with the RBF kernel, which successfully differentiates between various individuals.
Impact of Class Imbalance:

It is evident that classes with a greater number of samples, such as George W. Bush (146 instances) and Colin Powell (84 instances), display superior performance metrics. The model learns to recognize these faces more reliably due to the larger volume of training data available. Conversely, classes like Ariel Sharon (17 instances) and Junichiro Koizumi (16 instances) show lower performance, suggesting that the model struggles to generalize well with limited data.
Precision and Recall Balance:

For several classes, precision and recall are relatively balanced. For example, George W. Bush exhibits a precision of 0.90 and a recall of 0.86, indicating that the model effectively identifies this class while minimizing false positives. However, for Hugo Chavez, although precision is perfect at 1.00, the recall is lower at 0.63, which means the model misses some actual instances while not misclassifying any.
Robust F1-Scores:

The F1-scores for most classes are commendable, especially for George W. Bush (0.88) and Colin Powell (0.81). These scores demonstrate a solid balance between precision and recall, crucial for evaluating classification models effectively.
Generalization Across Classes:

The macro average F1-score of 0.79 indicates a reasonable consistency across different classes, while the weighted average F1-score of 0.82 shows that the performance is boosted by classes with more data. This suggests that while the model performs well overall, improvements can be made in handling classes with fewer instances.
Conclusion:
In summary, the model exhibits strong performance, particularly for classes with a higher number of samples. Nonetheless, it faces challenges with less represented classes, where precision and recall could be better optimized. Implementing strategies to address class imbalance, such as oversampling smaller classes or employing class-weighted SVMs, could enhance the model's effectiveness and improve its generalization across all categories.


